### PR TITLE
P-256 Optimizations for ARMv8 (AArch64)

### PR DESCRIPTION
**Pull request by nebeid**
### Issues:
* CryptoAlg-536: import OpenSSL ARMv8 assembly functions for P-256 ([PR#52](https://github.com/awslabs/aws-lc/pull/52))
* CryptoAlg-556: implement the 1st of 3 missing assembly functions in the imported code ([PR#52] (https://github.com/awslabs/aws-lc/pull/52))
* CryptoAlg-573: implement the 2nd of 3 missing assembly functions ([PR#56](https://github.com/awslabs/aws-lc/pull/56) )
* CryptoAlg-574: implement the 3rd of 3 missing assembly functions ([PR#57](https://github.com/awslabs/aws-lc/pull/57))
* CryptoAlg-591: Remove duplicate C code that was used during gradual development of ARMv8 optimization ([PR#60](https://github.com/awslabs/aws-lc/pull/60))
* CryptoAlg-587: Introduce a new macro, OPENSSL_AARCH64_P256, to enable P-256 ARMv8 optimisations ([PR#61](https://github.com/awslabs/aws-lc/pull/61))
* CryptoAlg-530: Merge changed files directly, delete the patch, create a new ci test and update `BUILDING.md`, this PR ([PR#62](https://github.com/awslabs/aws-lc/pull/62))

### Description of changes: 
The first 5 commits in this PR are cherry-picked from branch [p256-armv8](https://github.com/awslabs/aws-lc/tree/p256-armv8), where they were changes to a `quilt` patch. Subsequent commits committed the changed files directly and deleted the patch.

The changes in this PR introduce ARMv8 (AArch64) assembly optimizations for P-256 into AWS-LC. The goal is bringing the P-256 performance on ARMv8 at par with x86_64. The ARMv8 assembly code is taken from OpenSSL 1.1.1's [`openss/crypto/ec/asm/ecp_nistz256-armv8.pl`](https://github.com/openssl/openssl/blob/46a9ee8c796c8b5f8d95290676119b4f3d72be91/crypto/ec/asm/ecp_nistz256-armv8.pl)(at [this commit](https://github.com/openssl/openssl/commit/46a9ee8c796c8b5f8d95290676119b4f3d72be91)).

The goal is achieved by reusing the AWS-LC P-256 implementation in `p256-x86_64.c` (nistz256) with ARMv8. This is possible because the assembly functions found in `crypto/fipsmodule/ec/asm/p256-x86_64-asm.pl` required to support that code have their analogous functions in the imported OpenSSL ARMv8 Perl assembly file. 

Specifically, `ecp_nistz256-armv8.pl` was imported with the following modifications:
* Renamed to `p256-armv8-asm.pl`
* Modified the location of `arm-xlate.pl` and `arm_arch.h`
* Replaced the `scatter-gather subroutines` with `select subroutines`. The `select subroutines` are implemented for ARMv8 similarly to their x86_64 counterparts, `ecp_nistz256_select_w5` and `ecp_nistz256_select_w7`.
* Another assembly function called by the C code in `p256-x86_64.c` is `beeu_mod_inverse_vartime()`. This function was implemented for AArch64 in `p256_beeu-armv8-asm.pl` similarly to its implementation in `p256_beeu-x86_64-asm.pl`.

A new macro `OPENSSL_AARCH64_P256` is introduced to control the enabling of the ARMv8 P-256 optimizations. A new ci test is created to enable it and test the new code.

The files containing `p256-x86_64` in their name were renamed to, `p256-nistz` since the functions and tests defined in them are hereby running on ARMv8 as well, if enabled.

### Call-outs:
* This change focuses on AArch64 (64-bit architecture of ARMv8). It does not support ARMv4 or ARMv7.
* The assembly functions added contain _pointer authentication_ instructions that match the rest of the imported code. In subsequent PRs, pointer authentication (PAuth) and Branch Target Identification (BTI), will be modified/added to the assembly code introduced in this PR. They will conform to [this commit to BoringSSL](https://boringssl-review.googlesource.com/c/boringssl/+/42084) which is merged in `main`.
* ~The files containing `p256-x86_64` in their name should be ideally renamed to, e.g., `p256-nistz256` since the functions and tests defined in them are hereby running on ARMv8 as well, if enabled.~
* The changes described above were introduced gradually in the code base, hence the code duplication from `p256-x86_64.c` into `p256.c`, which was subsequently removed when all assembly functions called by that code were implemented for ARMv8.

### Testing:
Tested on ARMv8 and x86_64 platforms by running `cmake` as follows:
```
cmake -DOPENSSL_AARCH64_P256=1 -DCMAKE_BUILD_TYPE=Release -GNinja ..
ninja
./tool/bssl speed -filter "ECDH P-256"
./tool/bssl speed -filter "ECDSA P-256"
```
The speed tests report a significant difference based on whether `-DOPENSSL_AARCH64_P256=1` or `=0`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/ci/run_posix_tests.sh [hunk 1]
```diff
@@ -16,6 +16,9 @@ build_and_test -DOPENSSL_SMALL=1 -DCMAKE_BUILD_TYPE=Release
 echo "Testing AWS-LC in no asm mode."
 build_and_test -DOPENSSL_NO_ASM=1 -DCMAKE_BUILD_TYPE=Release
 
+echo "Testing AWS-LC with AArch64 P-256 optimizations."
+build_and_test -DOPENSSL_AARCH64_P256=1 -DCMAKE_BUILD_TYPE=Release
+
```
Comment (bryce-shang) [inline tests/ci/run_posix_tests.sh:21]: The flag `OPENSSL_AARCH64_P256` is arm specific. But this test will also be executed in x86. If this is the purpose, can the flag be renamed?

**Review by bryce-shang — COMMENTED** [review]
Is the patch file content in `nebeid:p256-armv8-quilt` branch same as that in `p256-armv8` branch? It's safe to have the two branches checked out and use `diff` to compare the patch file.

File: BUILDING.md [hunk 2]
```diff
@@ -85,6 +85,12 @@ remove some code that is especially large.
...
 
+### Building for ARMv8 (AArch64)
+In order to enable AArch64 optimizations for the NIST curve P-256, run `cmake`
+as follows:
+
+    cmake -DOPENSSL_AARCH64_P256=1 -DCMAKE_BUILD_TYPE=Release -GNinja ..
+
 ### Building for Android
 
 It's possible to build BoringSSL with the Android NDK using CMake. Recent
```
Comment (bryce-shang) [inline BUILDING.md:96]: Android is another ARM platform. Can the flag `OPENSSL_AARCH64_P256` be enabled in Android (which is not supported in CI)?

**Review by bryce-shang — COMMENTED** [review]
Nit: is the default strategy (squash) still preferred when merging this PR? If so, the cherry-picked commits will not show up in git history. This is my minor concern.

**Comment by nebeid** [comment]
> Nit: is the default strategy (squash) still preferred when merging this PR? If so, the cherry-picked commits will not show up in git history. This is my minor concern.
It will go in as separate commits, as per a quick discussion with Daryl.

**Comment by nebeid** [comment]
> Is the patch file content in `nebeid:p256-armv8-quilt` branch same as that in `p256-armv8` branch? It's better(safer) to have the two branches checked out and use `diff` to compare the patch file.

```
P256_ARMV8=p256_armv8
P256_ARMV8_QUILT=p256_armv8_quilt

# clone both branches in order to compare the patch
mkdir -p ${P256_ARMV8} && git clone --single-branch --branch p256-armv8 https://github.com/awslabs/aws-lc.git ${P256_ARMV8}
mkdir -p ${P256_ARMV8_QUILT} && git clone --single-branch --branch p256-armv8-quilt https://github.com/nebeid/aws-lc.git ${P256_ARMV8_QUILT}

# checkout the commit that is right before modifying the patch
cd ${P256_ARMV8_QUILT} && git checkout -b ca587-patch b150abe1c3f3ed2184aa06b439e6216dd9ea962f && cd ..

diff --brief ${P256_ARMV8}/patches/0013-p256-armv8-optimizations \
     ${P256_ARMV8_QUILT}/patches/0013-p256-armv8-optimizations
```
Result: no output => same patch

File: tests/ci/run_posix_tests.sh (hunk 1 above)
Comment (bryce-shang) [reply tests/ci/run_posix_tests.sh:21]: Any updates?

**Comment by nebeid** [comment]
Updated PR after [the last commit](https://github.com/awslabs/aws-lc/pull/62/commits/ca0ce9e67507049ae7f2dae96f5db9b618ef02b5)

File: tests/ci/run_posix_tests.sh (hunk 1 above)
Comment (nebeid) [reply tests/ci/run_posix_tests.sh:21]: That is a correct observation, Bryce. 
* My goal for now was to show that whether the switch was enabled or disabled, the tests run on both platforms.
* I prefer to keep AARCH64 in the flag name as it only pertains to the code run on that platform.
* In subsequent PR, the flag will likely be enabled in the code and not as a build flag (as per discussion with Daryl).

**Review by bryce-shang — APPROVED** [review]

**Review by torben-hansen — APPROVED** [review]
Nice

**Pull request merged**